### PR TITLE
Week 4 Complete

### PR DESCRIPTION
**Pull request by balderasdiana**
Screen Shot 2020-06-25 at 12 59 22 PM

Screen Shot 2020-06-25 at 12 59 22 PM

Tests Passed For Week 4

### CODE REVIEW

File: walkthroughs/week-4-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 1]
```diff
@@ -15,9 +15,80 @@
...
 
 public final class FindMeetingQuery {
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    
+    ArrayList<TimeRange> busyTimeRanges = new ArrayList();
+    ArrayList<TimeRange> freeTimeRanges = new ArrayList();
+
+    // No attendees, free the whole day
+    if(request.getAttendees().isEmpty()){
```
Comment (qidichen) [inline walkthroughs/week-4-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:28]: nit: indentation for parentheses
`if (` and `) {`

File: walkthroughs/week-4-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 2]
```diff
@@ -15,9 +15,80 @@
...
+    } else {
+        // Not busy, free the whole day
+        freeTimeRanges.add(TimeRange.fromStartEnd(TimeRange.START_OF_DAY, TimeRange.END_OF_DAY, true));
+    }
+
+    return freeTimeRanges;
   }
+
+
+    private ArrayList<TimeRange> getTimesRequestedAttendeesAreBusy(Collection<Event> events, MeetingRequest request){
```
Comment (qidichen) [inline walkthroughs/week-4-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:57]: Better to use interface for return type
`List< TimeRange >`

File: walkthroughs/week-4-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 3]
```diff
@@ -15,9 +15,80 @@
...
+            // Add busy time, when attending an event
+            if(!Collections.disjoint(event.getAttendees(), request.getAttendees())){
+                busyTimes.add(event.getWhen());
+            }
+        }
+
+        return busyTimes;
+    }
+
+    private void findFreeTimeBeforeFirstMeeting(ArrayList<TimeRange> busyTimeRanges, ArrayList<TimeRange> freeTimeRanges, MeetingRequest request){
```
Comment (qidichen) [inline walkthroughs/week-4-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:69]: For parameters, also better to use interface:  `List<TimeRange>`, which would give better flexibility.

File: walkthroughs/week-4-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 4]
```diff
@@ -15,9 +15,80 @@
...
+        if((busyTimeRanges.get(0).start() - TimeRange.START_OF_DAY) >= request.getDuration()){
+            freeTimeRanges.add(TimeRange.fromStartEnd(TimeRange.START_OF_DAY, busyTimeRanges.get(0).start(), false));
+        }
+    }
+
+    private void findFreeTimeInBetweenMeetings(ArrayList<TimeRange> busyTimeRanges, ArrayList<TimeRange> freeTimeRanges, MeetingRequest request){
+        for(int i = 0; i < busyTimeRanges.size()-1; i++){
```
Comment (qidichen) [inline walkthroughs/week-4-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:76]:
```suggestion
        for (int i = 0; i < busyTimeRanges.size() - 1; i++) {
```

**Review by qidichen — APPROVED** [review]
Nice

**Comment by gitmnk** [comment]
Great problem solving logic! Good job Diana..

**Pull request merged**